pyogrio/raw.py

```diff
@@ -559,3 +559,116 @@ def write(
+
+    # if dtypes is given, remove it from kwargs (dtypes is included in meta returned by
+    # read, and it is convenient to pass meta directly into write for round trip tests)
+    kwargs.pop("dtypes", None)
+    path = vsi_path(str(path))
+
+    if driver is None:
+        driver = detect_write_driver(path)
+
+    # verify that driver supports writing
+    if not ogr_driver_supports_write(driver):
+        raise DataSourceError(
+            f"{driver} does not support write functionality in GDAL "
+            f"{get_gdal_version_string()}"
+        )
+
+    # prevent segfault from: https://github.com/OSGeo/gdal/issues/5739
+    if append and driver == "FlatGeobuf" and get_gdal_version() <= (3, 5, 0):
+        raise RuntimeError(
+            "append to FlatGeobuf is not supported for GDAL <= 3.5.0 due to segfault"
+        )
+
+    if metadata is not None:
+        if layer_metadata is not None:
+            raise ValueError("Cannot pass both metadata and layer_metadata")
+        layer_metadata = metadata
+
+    # validate metadata types
+    for metadata in [dataset_metadata, layer_metadata]:
+        if metadata is not None:
+            for k, v in metadata.items():
+                if not isinstance(k, str):
+                    raise ValueError(f"metadata key {k} must be a string")
+
+                if not isinstance(v, str):
+                    raise ValueError(f"metadata value {v} must be a string")
+
+    if arrow_obj is not None and promote_to_multi is None:
+        promote_to_multi = (
+            geometry_type.startswith("Multi")
+            and driver in DRIVERS_NO_MIXED_SINGLE_MULTI
+        )
+
+    # TODO: does GDAL infer CRS automatically from geometry metadata?
+    if arrow_obj is not None and crs is None:
+        warnings.warn(
+            "'crs' was not provided.  The output dataset will not have "
+            "projection information defined and may not be usable in other "
+            "systems."
+        )
+
+    # preprocess kwargs and split in dataset and layer creation options
+    dataset_kwargs = _preprocess_options_key_value(dataset_options or {})
+    layer_kwargs = _preprocess_options_key_value(layer_options or {})
+    if kwargs:
+        kwargs = _preprocess_options_key_value(kwargs)
+        dataset_option_names = _parse_options_names(
+            _get_driver_metadata_item(driver, "DMD_CREATIONOPTIONLIST")
+        )
+        layer_option_names = _parse_options_names(
+            _get_driver_metadata_item(driver, "DS_LAYER_CREATIONOPTIONLIST")
+        )
+        for k, v in kwargs.items():
+            if k in dataset_option_names:
+                dataset_kwargs[k] = v
+            elif k in layer_option_names:
+                layer_kwargs[k] = v
+            else:
+                raise ValueError(f"unrecognized option '{k}' for driver '{driver}'")
+
+    # TODO: pass these as dataset kwargs?
+    # promote_to_multi=promote_to_multi,
+    # nan_as_null=nan_as_null,
```

Comment:
I suppose those can't be supported with Arrow writing. For `promote_to_multi`, we actually call a GDAL function on each individual geometry while looping for writing row by row. So with the batch writing of `write_arrow`, this would need to be done ahead of writing (and so for now I would leave that up to the user), or it would need to be done by GDAL under the hood (but currently there is no option for that).

For `nan_as_null`, if you already start from Arrow data, there is less need for this option, because you can already have proper nulls then. It was specifically added for numpy-based systems that use NaN as the null indicator.